From c989170eb01b503d5abc005d3312884598397d7a Mon Sep 17 00:00:00 2001
From: Gerrit Pape <pape@smarden.org>
Date: Wed, 19 Sep 2007 09:04:31 +0000
Subject: [PATCH] dash, test: handle arithmetic argument NULL as Not A Number

The test builtin is silent when using '' as arithmetic argument, but
prints a diagnostic error message when using e.g. 'foo'

 $ dash -c 'test 1 -eq "foo"'
 test: 1: foo: bad number
 $ dash -c 'test 1 -eq ""'
 $

This commit makes dash print the message in both cases.

Patch is from Oleg Verych <olecom@flower.upol.cz>, posted through
 http://bugs.debian.org/431320
---
 src/bltin/test.c |   15 ++++++++-------
 1 files changed, 8 insertions(+), 7 deletions(-)

diff --git a/src/bltin/test.c b/src/bltin/test.c
index f16c819..6bc22fd 100644
--- a/src/bltin/test.c
+++ b/src/bltin/test.c
@@ -409,13 +409,14 @@ getn(const char *s)
 	if (errno != 0)
 	      error("%s: out of range", s);
 
-	while (isspace((unsigned char)*p))
-	      p++;
-	
-	if (*p)
-	      error("%s: bad number", s);
-
-	return (int) r;
+       if (p != s) {
+               while (isspace((unsigned char)*p))
+                       p++;
+
+               if (!*p)
+                       return (int) r;
+       }
+       error("`%s': bad number", s);
 }
 
 static int
-- 
1.5.3.1

